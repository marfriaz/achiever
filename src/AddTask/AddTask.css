.AddTask {
  color: #dedede;
}

.AddTask h2 {
  margin-left: 16px;
}
