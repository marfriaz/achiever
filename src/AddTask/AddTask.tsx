import React, { useContext, useState } from "react";
import { useHistory } from "react-router-dom";
import AchieverForm from "../AchieverForm/AchieverForm";
import ApiContext from "../ApiContext";
import config from "../config";
import "./AddTask.css";

export default function AddTask() {
  interface stateInterface {
    name: string;
    details: string;
    goal_id: string;
  }

  const [state, setState] = useState<stateInterface>({
    name: "",
    details: "",
    goal_id: "1",
  });

  const history = useHistory();

  const { addTask, goals } = useContext(ApiContext)!;

  const handleSubmit = (event: React.ChangeEvent<HTMLFormElement>) => {
    event.preventDefault();
    console.log(state);
    const newTask = {
      name: state!.name,
      details: state!.details,
      goal_id: state!.goal_id,
      date_created: new Date(),
    };
    fetch(`${config.API_ENDPOINT}/api/tasks`, {
      method: "POST",
      headers: {
        "content-type": "application/json",
      },
      body: JSON.stringify(newTask),
    })
      .then((res) => {
        if (!res.ok) {
          return res.json().then((e) => Promise.reject(e));
        }
        return res.json();
      })
      .then((task) => {
        console.log("Yo task,", task);
        addTask(task);
        history.push(`/task/${task.id}`);
      })
      .catch((error) => {
        console.error({ error });
      });
  };

  // const handleGoalOptions = (goals: any) => {
  //   {
  //     if (goals === undefined) {
  //       return null;
  //     } else {
  //       return goals.map((goal: any) => (
  //         <option key={goal.id} value={goal.id}>
  //           {goal.name}
  //         </option>
  //       ));
  //     }
  //   }
  // };
  console.log(state.details);
  return (
    <section className="AddTask">
      <h2>Create a task</h2>
      <AchieverForm onSubmit={handleSubmit}>
        <div className="field">
          <label htmlFor="task-name-input">Name</label>
          <input
            type="text"
            id="task-name-input"
            name="task-name"
            onChange={(event: React.FormEvent) => {
              setState({
                name: (event.target as any).value,
                details: state.details,
                goal_id: state.goal_id,
              });
            }}
          />
        </div>
        <div className="field">
          <label htmlFor="task-content-input">Content</label>
          <textarea
            id="task-content-input"
            name="task-content"
            onChange={(event: React.FormEvent) => {
              setState({
                name: state.name,
                details: (event.target as any).value,
                goal_id: state.goal_id,
              });
            }}
          />
        </div>
        <div className="field">
          <label htmlFor="task-goal-select">Goal</label>
          <select
            id="task-goal-select"
            name="task-goal-id"
            onChange={(event: React.FormEvent) => {
              setState({
                name: state.name,
                details: state.details,
                goal_id: (event.target as any).value,
              });
            }}
          >
            {/* <option value={null}>...</option> */}

            {goals!.map((goal: any) => (
              <option key={goal.id} value={goal.id}>
                {goal.name}
              </option>
            ))}
          </select>
        </div>
        <div className="buttons">
          <button type="submit">Add task</button>
        </div>
      </AchieverForm>
    </section>
  );
}
