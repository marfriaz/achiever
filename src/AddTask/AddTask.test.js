import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import AddTask from "./AddTask";

describe(`AddTask component`, () => {
  const stubGoals = [
    {
      id: "b0715efe-ffaf-11e8-8eb2-f2801f1b9fd1",
      name: "Important",
    },
    {
      id: "b07161a6-ffaf-11e8-8eb2-f2801f1b9fd1",
      name: "Super",
    },
    {
      id: "b07162f0-ffaf-11e8-8eb2-f2801f1b9fd1",
      name: "Spangley",
    },
  ];

  it("renders the complete form", () => {
    const wrapper = shallow(<AddTask />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  // enzyme doesn't support React.createContext
  it.skip("renders the select options from goals", () => {
    const context = { goals: stubGoals };
    const select = shallow(<AddTask />, context).find("#task-goal-select");
    expect(toJson(select)).toMatchSnapshot();
  });
});
