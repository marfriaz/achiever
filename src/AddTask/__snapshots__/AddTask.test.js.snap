// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTask component renders the complete form 1`] = `
<section
  className="AddTask"
>
  <h2>
    Create a task
  </h2>
  <AchieverForm>
    <div
      className="field"
    >
      <label
        htmlFor="task-name-input"
      >
        Name
      </label>
      <input
        id="task-name-input"
        type="text"
      />
    </div>
    <div
      className="field"
    >
      <label
        htmlFor="task-content-input"
      >
        Content
      </label>
      <textarea
        id="task-content-input"
      />
    </div>
    <div
      className="field"
    >
      <label
        htmlFor="task-goal-select"
      >
        Goal
      </label>
      <select
        id="task-goal-select"
      >
        <option
          value={null}
        >
          ...
        </option>
      </select>
    </div>
    <div
      className="buttons"
    >
      <button
        type="submit"
      >
        Add task
      </button>
    </div>
  </AchieverForm>
</section>
`;

exports[`AddTask component renders the select options from goals 1`] = `
<select
  id="task-goal-select"
>
  <option
    value={null}
  >
    ...
  </option>
  <option
    key="b0715efe-ffaf-11e8-8eb2-f2801f1b9fd1"
    value="b0715efe-ffaf-11e8-8eb2-f2801f1b9fd1"
  >
    Important
  </option>
  <option
    key="b07161a6-ffaf-11e8-8eb2-f2801f1b9fd1"
    value="b07161a6-ffaf-11e8-8eb2-f2801f1b9fd1"
  >
    Super
  </option>
  <option
    key="b07162f0-ffaf-11e8-8eb2-f2801f1b9fd1"
    value="b07162f0-ffaf-11e8-8eb2-f2801f1b9fd1"
  >
    Spangley
  </option>
</select>
`;
